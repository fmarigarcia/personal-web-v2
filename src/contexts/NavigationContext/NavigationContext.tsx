import React, { createContext, useContext, useState } from 'react';
import type { ReactNode } from 'react';
import { SECTIONS } from '@utils/constants';

// Types
interface NavigationContextType {
    currentSection: string;
    setCurrentSection: (section: string) => void;
    isScrolling: boolean;
    setIsScrolling: (scrolling: boolean) => void;
}

// Create context
const NavigationContext = createContext<NavigationContextType | undefined>(
    undefined
);

// Provider component
interface NavigationProviderProps {
    children: ReactNode;
}

export const NavigationProvider: React.FC<NavigationProviderProps> = ({
    children,
}) => {
    const [currentSection, setCurrentSection] = useState<string>(SECTIONS.HERO);
    const [isScrolling, setIsScrolling] = useState<boolean>(false);

    const value: NavigationContextType = {
        currentSection,
        setCurrentSection,
        isScrolling,
        setIsScrolling,
    };

    return (
        <NavigationContext.Provider value={value}>
            {children}
        </NavigationContext.Provider>
    );
};

// Hook to use the context
export const useNavigation = (): NavigationContextType => {
    const context = useContext(NavigationContext);

    if (context === undefined) {
        throw new Error(
            'useNavigation must be used within a NavigationProvider'
        );
    }

    return context;
};
