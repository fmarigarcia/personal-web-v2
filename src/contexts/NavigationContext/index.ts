export {
    NavigationContext,
    type NavigationContextType,
} from './NavigationContext';
export { NavigationProvider } from './NavigationProvider';
export { useNavigation } from './useNavigation';
