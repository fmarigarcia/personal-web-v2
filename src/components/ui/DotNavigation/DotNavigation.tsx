import React from 'react';
import { useTranslation } from 'react-i18next';
import { useNavigation } from '@contexts/NavigationContext';
import { useSmoothScroll } from '@hooks/useSmoothScroll';
import { NAVIGATION_ITEMS } from '@utils/constants';

interface DotNavigationProps {
    className?: string;
}

export const DotNavigation: React.FC<DotNavigationProps> = ({
    className = '',
}) => {
    const { t } = useTranslation();
    const { currentSection } = useNavigation();
    const { scrollToElement } = useSmoothScroll();

    const handleDotClick = (sectionId: string) => {
        scrollToElement(sectionId);
    };

    return (
        <nav
            className={`fixed right-6 top-1/2 -translate-y-1/2 z-40 ${className}`}
            aria-label="Section navigation"
            role="navigation"
        >
            <ul className="space-y-3">
                {NAVIGATION_ITEMS.map((item) => (
                    <li key={item.id}>
                        <button
                            onClick={() => handleDotClick(item.id)}
                            aria-label={`Go to ${t(item.labelKey)} section`}
                            className={`
                                relative block w-3 h-3 rounded-full border-2 
                                transition-all duration-300 ease-in-out
                                hover:scale-125 focus:outline-none focus:ring-2 
                                focus:ring-blue-500 focus:ring-offset-2
                                ${
                                    currentSection === item.id
                                        ? 'bg-blue-600 border-blue-600 scale-110'
                                        : 'bg-transparent border-gray-400 hover:border-blue-500'
                                }
                            `}
                            type="button"
                        >
                            {/* Tooltip */}
                            <span
                                className="
                                    absolute right-full mr-3 px-2 py-1 
                                    bg-gray-900 text-white text-xs rounded 
                                    opacity-0 pointer-events-none 
                                    transition-opacity duration-200
                                    whitespace-nowrap top-1/2 -translate-y-1/2
                                    group-hover:opacity-100
                                "
                            >
                                {t(item.labelKey)}
                            </span>
                        </button>
                    </li>
                ))}
            </ul>
        </nav>
    );
};

export default DotNavigation;
