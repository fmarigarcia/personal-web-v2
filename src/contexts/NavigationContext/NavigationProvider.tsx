import React, { useState, useCallback, useRef } from 'react';
import type { ReactNode } from 'react';
import { SECTIONS } from '@utils/constants';
import { NavigationContext } from './NavigationContext';
import type { NavigationContextType } from './NavigationContext';

interface NavigationProviderProps {
    children: ReactNode;
}

export const NavigationProvider: React.FC<NavigationProviderProps> = ({
    children,
}) => {
    const [currentSection, setCurrentSection] = useState<string>(SECTIONS.HERO);
    const [isScrolling, setIsScrolling] = useState<boolean>(false);

    // Use a ref to store the navigation function to avoid infinite loops
    const navigateRef = useRef<((sectionId: string) => void) | undefined>(
        undefined
    );

    // Stable navigation function that calls the current ref value
    const navigateToSection = useCallback((sectionId: string) => {
        if (navigateRef.current) {
            navigateRef.current(sectionId);
        }
    }, []);

    // Stable setter function that updates the ref
    const setNavigateToSection = useCallback(
        (fn: (sectionId: string) => void) => {
            navigateRef.current = fn;
        },
        []
    );

    const value: NavigationContextType = {
        currentSection,
        setCurrentSection,
        isScrolling,
        setIsScrolling,
        navigateToSection,
        setNavigateToSection,
    };

    return (
        <NavigationContext.Provider value={value}>
            {children}
        </NavigationContext.Provider>
    );
};
