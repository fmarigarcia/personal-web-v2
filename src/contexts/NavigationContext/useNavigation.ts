import { useContext } from 'react';
import { NavigationContext } from './NavigationContext';
import type { NavigationContextType } from './NavigationContext';

// Hook to use the context
export const useNavigation = (): NavigationContextType => {
    const context = useContext(NavigationContext);

    if (context === undefined) {
        throw new Error(
            'useNavigation must be used within a NavigationProvider'
        );
    }

    return context;
};
