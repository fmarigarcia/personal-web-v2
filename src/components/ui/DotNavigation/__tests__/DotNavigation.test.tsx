import { render, screen, fireEvent } from '@testing-library/react';
import { DotNavigation } from '../DotNavigation';
import { NavigationProvider } from '@contexts/NavigationContext';

// Mock i18next
jest.mock('react-i18next', () => ({
    useTranslation: () => ({
        t: (key: string) => {
            const translations: Record<string, string> = {
                'nav.home': 'Home',
                'nav.about': 'About',
                'nav.experience': 'Experience',
                'nav.projects': 'Projects',
                'nav.contact': 'Contact',
            };
            return translations[key] || key;
        },
    }),
}));

// Mock hooks
const mockScrollToElement = jest.fn();
jest.mock('@hooks/useSmoothScroll', () => ({
    useSmoothScroll: () => ({
        scrollToElement: mockScrollToElement,
    }),
}));

describe('DotNavigation', () => {
    beforeEach(() => {
        mockScrollToElement.mockClear();
    });

    const renderDotNavigation = (props = {}) => {
        return render(
            <NavigationProvider>
                <DotNavigation {...props} />
            </NavigationProvider>
        );
    };

    it('should render navigation dots for all sections', () => {
        renderDotNavigation();

        // Should have 5 dots (one for each section)
        const dots = screen.getAllByRole('button');
        expect(dots).toHaveLength(5);

        // Check aria-labels contain section names
        expect(
            screen.getByRole('button', { name: /go to home section/i })
        ).toBeInTheDocument();
        expect(
            screen.getByRole('button', { name: /go to about section/i })
        ).toBeInTheDocument();
        expect(
            screen.getByRole('button', { name: /go to experience section/i })
        ).toBeInTheDocument();
        expect(
            screen.getByRole('button', { name: /go to projects section/i })
        ).toBeInTheDocument();
        expect(
            screen.getByRole('button', { name: /go to contact section/i })
        ).toBeInTheDocument();
    });

    it('should handle dot click navigation', () => {
        renderDotNavigation();

        const aboutDot = screen.getByRole('button', {
            name: /go to about section/i,
        });
        fireEvent.click(aboutDot);

        expect(mockScrollToElement).toHaveBeenCalledWith('about');
    });

    it('should highlight the current section', () => {
        renderDotNavigation();

        // Default current section should be 'hero'
        const homeDot = screen.getByRole('button', {
            name: /go to home section/i,
        });
        expect(homeDot).toHaveClass('bg-blue-600', 'border-blue-600');

        // Other dots should not be active
        const aboutDot = screen.getByRole('button', {
            name: /go to about section/i,
        });
        expect(aboutDot).toHaveClass('bg-transparent', 'border-gray-400');
    });

    it('should have proper accessibility attributes', () => {
        renderDotNavigation();

        const nav = screen.getByRole('navigation');
        expect(nav).toHaveAttribute('aria-label', 'Section navigation');

        const dots = screen.getAllByRole('button');
        dots.forEach((dot) => {
            expect(dot).toHaveAttribute('aria-label');
            expect(dot).toHaveAttribute('type', 'button');
        });
    });

    it('should apply custom className', () => {
        const customClass = 'custom-dot-nav';
        renderDotNavigation({ className: customClass });

        const nav = screen.getByRole('navigation');
        expect(nav).toHaveClass(customClass);
    });

    it('should be positioned fixed on the right side', () => {
        renderDotNavigation();

        const nav = screen.getByRole('navigation');
        expect(nav).toHaveClass(
            'fixed',
            'right-6',
            'top-1/2',
            '-translate-y-1/2',
            'z-40'
        );
    });

    it('should match snapshot', () => {
        const { container } = renderDotNavigation();
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should have hover and focus states', () => {
        renderDotNavigation();

        const dots = screen.getAllByRole('button');

        dots.forEach((dot) => {
            expect(dot).toHaveClass('hover:scale-125');
            expect(dot).toHaveClass(
                'focus:outline-none',
                'focus:ring-2',
                'focus:ring-blue-500'
            );
        });
    });
});
