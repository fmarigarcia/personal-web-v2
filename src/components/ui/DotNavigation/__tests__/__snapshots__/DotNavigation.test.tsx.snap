// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DotNavigation should match snapshot 1`] = `
<nav
  aria-label="Section navigation"
  class="fixed right-6 top-1/2 -translate-y-1/2 z-40 "
  role="navigation"
>
  <ul
    class="space-y-3"
  >
    <li>
      <button
        aria-label="Go to Home section"
        class="
                                relative block w-3 h-3 rounded-full border-2 
                                transition-all duration-300 ease-in-out
                                hover:scale-125 focus:outline-none focus:ring-2 
                                focus:ring-blue-500 focus:ring-offset-2
                                bg-blue-600 border-blue-600 scale-110
                            "
        type="button"
      >
        <span
          class="
 absolute right-full mr-3 px-2 py-1 
 bg-gray-900 text-white text-xs rounded 
 opacity-0 pointer-events-none 
 transition-opacity duration-200
 whitespace-nowrap top-1/2 -translate-y-1/2
 group-hover:opacity-100
 "
        >
          Home
        </span>
      </button>
    </li>
    <li>
      <button
        aria-label="Go to About section"
        class="
                                relative block w-3 h-3 rounded-full border-2 
                                transition-all duration-300 ease-in-out
                                hover:scale-125 focus:outline-none focus:ring-2 
                                focus:ring-blue-500 focus:ring-offset-2
                                bg-transparent border-gray-400 hover:border-blue-500
                            "
        type="button"
      >
        <span
          class="
 absolute right-full mr-3 px-2 py-1 
 bg-gray-900 text-white text-xs rounded 
 opacity-0 pointer-events-none 
 transition-opacity duration-200
 whitespace-nowrap top-1/2 -translate-y-1/2
 group-hover:opacity-100
 "
        >
          About
        </span>
      </button>
    </li>
    <li>
      <button
        aria-label="Go to Experience section"
        class="
                                relative block w-3 h-3 rounded-full border-2 
                                transition-all duration-300 ease-in-out
                                hover:scale-125 focus:outline-none focus:ring-2 
                                focus:ring-blue-500 focus:ring-offset-2
                                bg-transparent border-gray-400 hover:border-blue-500
                            "
        type="button"
      >
        <span
          class="
 absolute right-full mr-3 px-2 py-1 
 bg-gray-900 text-white text-xs rounded 
 opacity-0 pointer-events-none 
 transition-opacity duration-200
 whitespace-nowrap top-1/2 -translate-y-1/2
 group-hover:opacity-100
 "
        >
          Experience
        </span>
      </button>
    </li>
    <li>
      <button
        aria-label="Go to Projects section"
        class="
                                relative block w-3 h-3 rounded-full border-2 
                                transition-all duration-300 ease-in-out
                                hover:scale-125 focus:outline-none focus:ring-2 
                                focus:ring-blue-500 focus:ring-offset-2
                                bg-transparent border-gray-400 hover:border-blue-500
                            "
        type="button"
      >
        <span
          class="
 absolute right-full mr-3 px-2 py-1 
 bg-gray-900 text-white text-xs rounded 
 opacity-0 pointer-events-none 
 transition-opacity duration-200
 whitespace-nowrap top-1/2 -translate-y-1/2
 group-hover:opacity-100
 "
        >
          Projects
        </span>
      </button>
    </li>
    <li>
      <button
        aria-label="Go to Contact section"
        class="
                                relative block w-3 h-3 rounded-full border-2 
                                transition-all duration-300 ease-in-out
                                hover:scale-125 focus:outline-none focus:ring-2 
                                focus:ring-blue-500 focus:ring-offset-2
                                bg-transparent border-gray-400 hover:border-blue-500
                            "
        type="button"
      >
        <span
          class="
 absolute right-full mr-3 px-2 py-1 
 bg-gray-900 text-white text-xs rounded 
 opacity-0 pointer-events-none 
 transition-opacity duration-200
 whitespace-nowrap top-1/2 -translate-y-1/2
 group-hover:opacity-100
 "
        >
          Contact
        </span>
      </button>
    </li>
  </ul>
</nav>
`;
